File: tests/typesafety/python/everyday_higher_order_misuse.py
# expect: fails
# title: An extractor that returns the item name, rejected by the checker
# misuse-of: everyday_higher_order
from dataclasses import dataclass
from typing import NewType

Pence = NewType("Pence", int)


@dataclass(frozen=True)
class InvoiceLine:
    item: str
    pence: Pence
    gift: bool


def is_gift(line: InvoiceLine) -> bool:
    return line.gift


def label(line: InvoiceLine) -> str:
    return line.item


def gift_total(lines: list[InvoiceLine]) -> Pence:
    return Pence(sum(map(label, filter(is_gift, lines))))

File: tests/typesafety/typescript/everyday_higher_order_misuse.ts
// expect: fails

declare const penceBrand: unique symbol;
type Pence = number & { readonly [penceBrand]: true };

type InvoiceLine = { readonly item: string; readonly pence: Pence; readonly gift: boolean };

function isGift(line: InvoiceLine): boolean {
  return line.gift;
}

function label(line: InvoiceLine): string {
  return line.item;
}

export function giftTotal(lines: readonly InvoiceLine[]): number {
  return lines.filter(isGift).map(label).reduce((sum, pence) => sum + pence, 0);
}
